[SYSTEMML-906] [SYSTEMML-907] Update df first to df schema

deroneriksson opened the pull request:
Utilize DataFrame schema to determine matrix/frame characteristics
rather than calling first().

deroneriksson commented:
All tests pass here: https://sparktc.ibmcloud.com/jenkins/job/SystemML-OnDemand/205/console

deroneriksson commented:
cc @dusenberrymw

dusenberrymw commented on src/main/java/org/apache/sysml/api/mlcontext/MLContextUtil.java:
```diff
@@ -585,24 +587,24 @@ public static boolean isCSVLineAllNumbers(String line) {
+		}
+		for (StructField field : fields) {
+			DataType dataType = field.dataType();
+			if ((dataType != DataTypes.DoubleType) && (dataType != DataTypes.IntegerType)
```
We may need to add the `BigInt` type here as well.  At least from Python, when an RDD of integers is converted to a DataFrame with `toDF()`, the integers are set to type `BigInt`.  I'm testing the behavior of this in the current master to see if the conversion happens smoothly even with `BigInt`.

dusenberrymw replied:
Confirmed that DataFrames with a `BigInt` type (i.e. a bigint for the `__INDEX`) will get converted correctly on the master branch.  Let's just add an additional `BigInt` type here.

dusenberrymw commented on src/test/java/org/apache/sysml/test/integration/mlcontext/MLContextTest.java:
```diff
@@ -561,13 +561,13 @@ public void testDataFrameSumDMLDoublesWithIDColumn() {
-		fields.add(DataTypes.createStructField("C1", DataTypes.StringType, true));
-		fields.add(DataTypes.createStructField("C2", DataTypes.StringType, true));
-		fields.add(DataTypes.createStructField("C3", DataTypes.StringType, true));
+		fields.add(DataTypes.createStructField("__INDEX", DataTypes.DoubleType, true));
```
I think we should set to this to `IntegerType` -- perhaps even throw an error if the `__INDEX` is a double.

deroneriksson commented:
@dusenberrymw On the Java side, looks like the bigint gets converted to a long, so I added a check for DataTypes.LongType. Can be verified with:
```
import org.apache.spark.mllib.linalg._
val row1 = (1L, Vectors.dense(1.0,2.0,3.0))
val row2 = (2L, Vectors.dense(4.0,5.0,6.0))
val rdd = sc.parallelize(Array(row1, row2))
val df = rdd.toDF("__INDEX", "C1")
df.show

import org.apache.sysml.api.mlcontext._
val ml = new MLContext(sc)
val scr = ScriptFactory.dml("print(toString(M))").in("M", df)
ml.execute(scr)
```

I also updated the tests to use DataTypes.IntegerType.

dusenberrymw commented:
The following is now confirmed to work on PySpark:

```python
from systemml import MLContext, dml
from pyspark.mllib.linalg import Vectors

X = sc.parallelize([(1, Vectors.dense([3.4,4.3,5.3])), (2, Vectors.dense([5.2,6.5,7.0]))]).toDF(["__INDEX", "sample"])

ml = MLContext(sc)
script = """
s = sum(X)
"""
script = dml(script).input(X=X).output("s")
s = ml.execute(script).get("s")
```

deroneriksson commented:
Thank you for including the PySpark example. Very helpful if needed later.
